### DISCUSSION TITLE
How to change the parity on the drive?

### DISCUSSION THREAD

**Discussion by stanhausc** [Louisvdw/dbus-serialbattery#388] [Q&A] [answered]
Hello, I'm having trouble implementing a drive for a new BMS, it's returning BOLL instead of the frame, I think it must be the wrong parity, is it possible to change it (none to even)?

![image](https://user-images.githubusercontent.com/85182432/210412178-ab0613b7-ede6-4a7a-ad73-607bc501cf86.png)

**Comment by Louisvdw** [comment]
Most devices use a standard parity ever since dailup modem times, so I expect that is not your issue. 
Check how many bytes are returned and what the value is for each byte. Then compare that with what you expect for the BBBH which is 3 bytes and a short.
Also see https://docs.python.org/3/library/struct.html

**Comment by stanhausc** [comment]
This is an example to read all necessary data (voltage, current, soc, cell number is 1):

![image](https://user-images.githubusercontent.com/85182432/210624530-39b99faf-2bb3-4ee8-a97d-ea7db2c90fc0.png)

Frame sent:   command_geral = b"\x01\x03\x00\x00\x00\x05\x85\xC9"
Frame Received: 01 03 0A 61 F1 27 10 00 00 02 0A 0B 5A 8B D7

But what is returning is "FALSE", that's why the boolean error instead of receiving hexadecimal, the frames above I tested in a software, and in this software it was necessary to use "EVEN" parity, that's why my doubt on how to change it. @Louisvdw  think that's not it?

I also get logger.error(">>> ERROR: No reply - returning"), it was from there that I was in doubt about the parity.

![image](https://user-images.githubusercontent.com/85182432/210623792-66a1eb9c-3858-49f5-a024-52a664501dd3.png)

**Reply by Louisvdw** [reply]
IF you tested it, then it could very well be the parity.
See the 4th yellow paragraph for an example how to set the partity on the pyserial connection
https://pyserial.readthedocs.io/en/latest/shortintro.html#opening-serial-ports

**Reply by stanhausc** [reply]
Thanks @Louisvdw, it worked by changing the parity.

I have another question, this battery doesn't have cells, should I put cell_count = 1? I believe this information is missing for the data to appear on the screen.
![image](https://user-images.githubusercontent.com/85182432/211827613-daad0237-301a-4f6d-a61b-6966bab69c69.png)
![image](https://user-images.githubusercontent.com/85182432/211828179-8777a7a2-626d-4e4e-8f15-7d85c6986bac.png)

**Reply by Louisvdw** [reply]
There has to be cells connected before the BMS will work with the driver. So I assume you are saying that the BMS does not publish cells. In that case you don't populate the cells[] list, but you do need to set the cell_count to the real amount of cells in your battery. 

So there can be less cells in the cell list (normally if you don't have cell data) than what is set in cell_count.

**Reply by stanhausc** [reply]
Ok, I inserted a number of cells, can you tell me why it doesn't appear on the screen?
"""
def read_general(self):
         all_data = self.read_serial_data_fsonick(self.command_general)

id, function, length, voltage, current, nsei, soc, temperature = unpack_from(">BBBHHHHH", all_data)

         self.id = id
         self.function = function
         self.length = length
         self.voltage = voltage / 100
         self.current = current / 100
         self.soc = soc / 10
         self.temp1 = temperature / 100
         self.cell_count = int(15)
"""

![image](https://user-images.githubusercontent.com/85182432/211893936-fd4dc263-c3f7-4399-8753-8cd03f1a6a61.png)

![image](https://user-images.githubusercontent.com/85182432/211893778-aac444b8-2097-415a-b511-88344c5de1d5.png)

**Reply by Louisvdw** [reply]
in utils.py around line 11 change the logging level to DEBUG
>logger.setLevel(logging.DEBUG)

Then add logger.debug statements that will show you your values in the log file as they are read. You can use this to see what the values are. If there are no values they will not display.

**Reply by stanhausc** [reply, marked as answer]
@Louisvdw the values are taken by the script, however the BMS keeps connecting/disconnecting, and they don't appear on the screen.

![image](https://user-images.githubusercontent.com/85182432/214072115-ab144e11-5736-447d-a9a3-eeb53602d746.png)
![image](https://user-images.githubusercontent.com/85182432/214076127-fe813d0e-1f27-4433-a175-d155e66438d8.png)


I also modified a part of the utills.py file, so it started to return data (voltage, current, soc, temp), without that the data would not be sent

![image](https://user-images.githubusercontent.com/85182432/214077310-26753c82-d41f-491a-aaab-76ef58e7a7fb.png)

should I not populate the cells?

![image](https://user-images.githubusercontent.com/85182432/214078360-2a728120-7e5f-453f-b02c-044b5f217fa4.png)

**Comment by stanhausc** [comment]
It worked, thanks for the help @Louisvdw !!!!!!!

![image](https://user-images.githubusercontent.com/85182432/214651831-75e076db-dd3c-4cd5-83be-0be3c68015dc.png)


![image](https://user-images.githubusercontent.com/85182432/214651862-662010f8-b560-4e96-ae07-20883bac6ddf.png)

**Reply by Louisvdw** [reply]
Great!
